core/src/main/java/org/apache/iceberg/util/StructLikeMap.java

```diff
@@ -146,25 +145,19 @@ public R getValue() {
 
     @Override
     public int hashCode() {
-      int hashCode = getKey().hashCode();
-      if (getValue() != null) {
-        hashCode ^= getValue().hashCode();
-      }
-      return hashCode;
+      return inner.hashCode();
     }
 
     @Override
-    @SuppressWarnings("unchecked")
     public boolean equals(Object o) {
       if (this == o) {
         return true;
-      } else if (!(o instanceof StructLikeEntry)) {
+      } else if (o == null || getClass() != o.getClass()) {
         return false;
-      } else {
-        StructLikeEntry that = (StructLikeEntry<R>) o;
-        return Objects.equals(getKey(), that.getKey())
```

Comment:
We used `getKey()` which unwrapped the object and we were loosing proper equality.